torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -441,19 +443,100 @@ void ONNXSetDynamicInputShape(
-  for (size_t i = 0; i < outputs.size(); ++i) {
-    if (onnx_shape_inference) {
-      graph->outputs()[i]->setType(MergeInferredType(
-          TensorType::create(outputs[i]), graph->outputs()[i]->type()));
-    } else {
-      graph->outputs()[i]->inferTypeFrom(outputs[i]);
+  size_t outputs_index = 0;
+
+  PyObject* py_obj = unflatten(outputs, desc);
+
+  TORCH_INTERNAL_ASSERT(PyTuple_Check(py_obj));
+
+  for (size_t i = 0; i < PyTuple_GET_SIZE(py_obj); ++i) {
+    PyObject* elem = PyTuple_GET_ITEM(py_obj, i);
+
+    if (PyList_Check(elem)) {
+      size_t list_len = PyList_GET_SIZE(elem);
+      TORCH_INTERNAL_ASSERT(HasSequenceTypeOutput(graph->outputs()[i]->node()))
+      if (list_len > 0) {
+        auto& var =
+            reinterpret_cast<THPVariable*>(PyList_GET_ITEM(elem, 0))->cdata;
+        for (size_t j = 1; j < list_len; ++j) {
+          PyObject* list_elem = PyList_GET_ITEM(elem, j);
+          TORCH_INTERNAL_ASSERT(THPVariable_Check(list_elem));
+          auto& new_var = reinterpret_cast<THPVariable*>(list_elem)->cdata;
+          TORCH_CHECK(
+              var.scalar_type() == new_var.scalar_type(),
+              "Unsupported sequence type in model outputs. ONNX supports sequences of elements of the same data type.");
+        }
+        graph->outputs()[i]->setType(ListType::create(TensorType::create(var)));
+        outputs_index++;
+      }
+    } else if (PyTuple_Check(elem)) {
+      size_t tuple_len = PyTuple_GET_SIZE(elem);
+      if (tuple_len > 0) {
+        for (size_t j = 0; j < tuple_len; ++j) {
+          PyObject* tuple_elem = PyTuple_GET_ITEM(elem, j);
+          TORCH_INTERNAL_ASSERT(THPVariable_Check(tuple_elem));
+          auto& var = reinterpret_cast<THPVariable*>(tuple_elem)->cdata;
+          graph->outputs()[outputs_index + j]->setType(TensorType::create(var));
+        }
+        outputs_index += tuple_len;
+      }
+    } else if (THPVariable_Check(elem)) {
+      at::Tensor var = reinterpret_cast<THPVariable*>(elem)->cdata;
+      ONNXUpdateTypeFromTensor(
+          graph->outputs()[outputs_index], var, onnx_shape_inference);
+      outputs_index++;
+    } else { // Dict
+      // Support for dict data type is limited to fixed size dictionaries in
+      // ONNX.
+      // Dictionary values are unrolled and keys are not preserved.
+      TORCH_INTERNAL_ASSERT(PyDict_Check(elem));
+      auto unrolled_dict = py::reinterpret_borrow<py::list>(PyDict_Items(elem));
+      TORCH_INTERNAL_ASSERT(PyList_Check(unrolled_dict.ptr()));
+      for (size_t j = 0; j < unrolled_dict.size(); ++j) {
+        PyObject* tuple_elem = PyList_GET_ITEM(unrolled_dict.ptr(), j);
+        TORCH_INTERNAL_ASSERT(PyTuple_Check(tuple_elem));
+        TORCH_INTERNAL_ASSERT(PyTuple_GET_SIZE(tuple_elem) == 2);
+        auto& var =
+            reinterpret_cast<THPVariable*>(PyTuple_GET_ITEM(tuple_elem, 1))
+                ->cdata;
+        graph->outputs()[outputs_index + j]->setType(TensorType::create(var));
```

Comment:
should we use `ONNXUpdateTypeFromTensor`?

Reply:
As we will enable shape inference by default, do you think it makes sense to use ONNXUpdateTypeFromTensor or directly merge with MergeInferredType?
Is there a use-case to keep the conditional?